compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/gc/shenandoah/ShenandoahBarrierSet.java

```diff
@@ -0,0 +1,362 @@
+            case ReadNode readNode -> addReadNodeBarriers(readNode);
+            case WriteNode write -> addWriteBarriers(write, write.value(), null);
+            case LoweredAtomicReadAndWriteNode atomic -> {
+                if (useCASBarrier) {
+                    addWriteBarriers(atomic, atomic.getNewValue(), null);
+                    addReadNodeBarriers(atomic);
+                }
+            }
+            case AbstractCompareAndSwapNode cmpSwap -> {
+                if (useCASBarrier) {
+                    addWriteBarriers(cmpSwap, cmpSwap.getNewValue(), cmpSwap.getExpectedValue());
+                    if (cmpSwap instanceof ValueCompareAndSwapNode) {
+                        addReadNodeBarriers(cmpSwap);
+                    }
+                }
+            }
+            case ArrayRangeWrite ignored -> GraalError.unimplemented("ArrayRangeWrite is not used");
+            case null, default ->
+                GraalError.guarantee(n.getBarrierType() == BarrierType.NONE, "missed a node that requires a GC barrier: %s", n.getClass());
+        }
+    }
+
+    private void addWriteBarriers(FixedAccessNode node, ValueNode writtenValue, ValueNode expectedValue) {
+        BarrierType barrierType = node.getBarrierType();
+        switch (barrierType) {
+            case NONE:
+                // nothing to do
+                break;
+            case FIELD:
+            case ARRAY:
+            case UNKNOWN:
+            case POST_INIT_WRITE:
+            case AS_NO_KEEPALIVE_WRITE:
+                if (isObjectValue(writtenValue)) {
+                    StructuredGraph graph = node.graph();
+                    boolean init = node.getLocationIdentity().isInit();
+                    if (!init && barrierType != BarrierType.AS_NO_KEEPALIVE_WRITE && useSATBBarrier) {
+                        // The pre barrier does nothing if the value being read is null, so it can
+                        // be explicitly skipped when this is an initializing store.
+                        // No keep-alive means no need for the pre-barrier.
+                        addShenandoahSATBBarrier(node, node.getAddress(), expectedValue, graph);
+                    }
+                    if (!init && useCardBarrier && !StampTool.isPointerAlwaysNull(writtenValue)) {
+                        graph.addAfterFixed(node, graph.add(new ShenandoahCardBarrierNode(node.getAddress())));
+                    }
+                }
+                break;
+            default:
+                throw new GraalError("unexpected barrier type: " + barrierType);
+        }
+    }
+
+    private void addLoadReferenceBarrier(FixedAccessNode node, AddressNode address, BarrierType barrierType) {
+        GraalError.guarantee(node != null, "input value must not be null");
+        StructuredGraph graph = node.graph();
+        boolean narrow = node.stamp(NodeView.DEFAULT) instanceof NarrowOopStamp;
+        ValueNode uncompressed = maybeUncompressReference(node, narrow);
```

Comment:
It seems really important here that the `maybeUncompressReference` is done without unique because we don't want to reuse an existing uncompress node that is a usage of `node`. The implementation does this correctly, but you could make the requirement more explicit like this:
```
        Graph.Mark beforeUncompression = graph.getMark();
        ValueNode uncompressed = maybeUncompressReference(node, narrow);
        if (uncompressed != node) {
            GraalError.guarantee(graph.isNew(beforeUncompression, uncompressed), "we must not reuse an existing uncompress node");
        }
```